src/analysis/processing/qgsalgorithmunion.cpp

```diff
@@ -53,25 +53,23 @@ QgsProcessingAlgorithm *QgsUnionAlgorithm::createInstance() const
 void QgsUnionAlgorithm::initAlgorithm( const QVariantMap & )
 {
   addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
-  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "OVERLAY" ), QObject::tr( "Union layer" ) ) );
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "OVERLAY" ), QObject::tr( "Union layer" ), QList< int >(), QVariant(), true ) );
 
   addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Union" ) ) );
 }
 
-
 QVariantMap QgsUnionAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
 {
   std::unique_ptr< QgsFeatureSource > sourceA( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
   if ( !sourceA )
     throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
 
   std::unique_ptr< QgsFeatureSource > sourceB( parameterAsSource( parameters, QStringLiteral( "OVERLAY" ), context ) );
-  if ( !sourceB )
-    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "OVERLAY" ) ) );
+  // sourceB is optional so we do not throw an error if it is not a valid source
```

Comment:
I'd use something like:

    if ( parameters.contains( QStringLiteral( "OVERLAY" ) ) && !sourceB )
        // throw exception

This way if someone specifies the source, but it's not valid, they'll get the error

Reply:
I wanted to do something like that - but:
- the contains() approach does not work - processing GUI sends empty QString if the layer is not defined
- the param.toString().isEmpty() approach does not work either - when I choose a layer, it stores it in QgsProperty, so the toString() gives an empty string.
- at this point I gave up - we would need to partially duplicate the logic of QgsProcessingUtils::variantToSource() to cover various valid/invalid options. Or maybe to make the GUI not set a parameter as empty string if it is not set - but maybe that would break something I am not aware of...

Reply:
Is it sending an empty string or an invalid qvariant? If it's an empty string then this is a bug.

Reply:
Hmm I have checked again and it sends an invalid qvariant for OVERLAY parameter.

But then looking into the processing log, it says
```
Processing algorithm…
Algorithm 'Union' starting…
Input parameters:
{ 'INPUT' : 'MultiPolygon?crs=EPSG:3857&field=id_a:string&uid={c2163d44-911e-4bfd-b754-163055b23c20}', 'OUTPUT' : 'memory:', 'OVERLAY' : '' }

Execution completed in 0.05 seconds
Results:
{'OUTPUT': 'output_84d146ee_17fc_4418_80eb_81cda260168a'}
```
Which was probably the source of my confusion since says that OVERLAY is an empty string.

To add some more confusion, in processing history the OVERLAY parameter does not show up at all:
```
processing.run("native:union", {'INPUT':'/home/martin/qgis/git-master/python/plugins/processing/tests/testdata/custom/overlay0.geojson','OUTPUT':'memory:'})
```

Are missing parameter value and null qvariant parameter value are both acceptable ways of saying that a parameter is has not been specified?

Reply:
> But then looking into the processing log, it says

That's a bug too - should be None. QgsProcessingParameterFeatureSource::valueAsPythonString needs a check there to return "None" if value is not valid.

> To add some more confusion, in processing history the OVERLAY parameter does not show up at all:

That's another bug :) I think QgsProcessingAlgorithm::asPythonCommand should not be skipping invalid parameter values and instead be handing them off to valueAsPythonString to deal with.

Reply:
Fixes in #6964

Reply:
Are you able to re-introduce the exception now?

Reply:
Thanks for the fixes in #6964 - reintroduced the check and it work like a charm.